agent/cert-monitor/cert_monitor.go

```diff
@@ -0,0 +1,468 @@
+	// good way to make that happen and this behavior is benign enough
+	// that I am going to push off implementing it.
+
+	// the agent token has been updated so we must update our leaf cert watch.
+	// this cancels the current watches before setting up new ones
+	m.cancelWatches()
+
+	// recreate the chan for cache updates. This is a precautionary measure to ensure
+	// that we don't accidentally get notified for the new watches being setup before
+	// a blocking query in the cache returns and sends data to the old chan. In theory
+	// the code in agent/cache/watch.go should prevent this where we specifically check
+	// for context cancellation prior to sending the event. However we could cancel
+	// it after that check and finish setting up the new watches before getting the old
+	// events. Both the go routine scheduler and the OS thread scheduler would have to
+	// be acting up for this to happen. Regardless the way to ensure we don't get events
+	// for the old watches is to simply replace the chan we are expecting them from.
+	close(m.cacheUpdates)
+	m.cacheUpdates = make(chan cache.UpdateEvent, 10)
+
+	// restart watches - this will be done with the correct token
+	cancelWatches, err := m.setupCacheWatches(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to restart watches after agent token update: %w", err)
+	}
+	m.cancelWatches = cancelWatches
+	return nil
+}
+
+// handleFallback is used when the current TLS certificate has expired and the normal
+// updating mechanisms have failed to renew it quickly enough. This function will
+// use the configured fallback mechanism to retrieve a new cert and start monitoring
+// that one.
+func (m *CertMonitor) handleFallback(ctx context.Context) error {
+	m.logger.Warn("agent's client certificate has expired")
+	// Background because the context is mainly useful when the agent is first starting up.
+	reply, err := m.fallback(ctx)
+	if err != nil {
+		return fmt.Errorf("error when getting new agent certificate: %w", err)
+	}
+
+	return m.Update(reply)
+}
+
+// run is the private method to be spawn by the Start method for
+// executing the main monitoring loop.
+func (m *CertMonitor) run(ctx context.Context, exit chan struct{}) {
+	// The fallbackTimer is used to notify AFTER the agents
+	// leaf certificate has expired and where we need
+	// to fall back to the less secure RPC endpoint just like
+	// if the agent was starting up new.
+	//
+	// Check 10sec (fallback leeway duration) after cert
+	// expires. The agent cache should be handling the expiration
+	// and renew it before then.
+	//
+	// If there is no cert, AutoEncryptCertNotAfter returns
+	// a value in the past which immediately triggers the
+	// renew, but this case shouldn't happen because at
+	// this point, auto_encrypt was just being setup
+	// successfully.
+	calcFallbackInterval := func() time.Duration {
+		certExpiry := m.tlsConfigurator.AutoEncryptCertNotAfter()
+		return certExpiry.Sub(time.Now().Add(m.fallbackLeeway))
```

Comment:
I think this is actually checking 10 seconds **before** the expiry time.

For example:

- A `NotAfter` expiry at time `100` seconds.
- A `time.Now()` value of `50` seconds.
- A `m.fallbackLeeway` value of `10` seconds.

`100 - (50 + 10) = 40`, so we would wait `40` more seconds until `90` and then trigger the fallback timer, but we really want to wait `60` seconds until `110`.

Reply:
Yep, I think you are correct and this is another auto-encrypt bug to be fixed. I think it should be `NotAfter.Add(leeway).Sub(now)` which is basically the duration of time between when we want the chan to fire (expiration + leeway) and now.

Reply:
This changes the behavior of the main bug being fixed from attempting to use an expired certificate for 10s if we fail to update the auto-encrypt certificate long enough to indefinitely using an expired certificate.

Reply:
Fixed it.